File: companion/ios/ScreenshotToolCompanion/Network/FrameTransport.swift
import Foundation
import CoreVideo
import UIKit

class FrameTransport {
    private let client: WebSocketClient
    private var lastFrameTime: CFAbsoluteTime = 0
    private let minFrameInterval: CFAbsoluteTime

    init(client: WebSocketClient, fps: Int = 10) {
        self.client = client
        self.minFrameInterval = 1.0 / Double(fps)
    }

    /// Convert a CVPixelBuffer to JPEG and send via WebSocket to all connected desktops.
    /// Rate-limited to configured fps.
    func sendFrame(from pixelBuffer: CVPixelBuffer) {
        let now = CFAbsoluteTimeGetCurrent()
        guard now - lastFrameTime >= minFrameInterval else { return }
        lastFrameTime = now

        guard let jpegData = pixelBufferToJpeg(pixelBuffer, quality: 0.7) else { return }
        client.sendFrameToAll(jpegData)
    }

    /// Convert a CGImage to JPEG and send.
    func sendFrame(from cgImage: CGImage) {
        let now = CFAbsoluteTimeGetCurrent()
        guard now - lastFrameTime >= minFrameInterval else { return }
        lastFrameTime = now

        let uiImage = UIImage(cgImage: cgImage)
        guard let jpegData = uiImage.jpegData(compressionQuality: 0.7) else { return }
        client.sendFrameToAll(jpegData)
    }

    /// Convert a CGImage to PNG and send as screenshot result.
    func sendScreenshot(from cgImage: CGImage) {
        let uiImage = UIImage(cgImage: cgImage)
        guard let pngData = uiImage.pngData() else { return }
        client.sendScreenshotToAll(pngData)
    }

    private func pixelBufferToJpeg(_ pixelBuffer: CVPixelBuffer, quality: CGFloat) -> Data? {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext()

        guard let cgImage = context.createCGImage(ciImage, from: ciImage.extent) else {
            return nil
        }

        let uiImage = UIImage(cgImage: cgImage)
        return uiImage.jpegData(compressionQuality: quality)
    }
}

File: companion/android/app/src/main/java/com/pixyvibe/companion/network/FrameTransport.kt
package com.pixyvibe.companion.network

import android.graphics.Bitmap
import android.util.Base64
import java.io.ByteArrayOutputStream

/**
 * Converts Bitmap frames to base64 JPEG/PNG and sends via WebSocketClient.
 * Mirrors iOS FrameTransport.swift.
 */
class FrameTransport(
    private val client: WebSocketClient,
    fps: Int = 10,
) {
    private val minFrameIntervalMs = 1000L / fps
    private var lastFrameTimeMs = 0L

    /**
     * Compress bitmap to JPEG and send as a frame to all connected desktops.
     * Rate-limited to configured fps.
     */
    fun sendFrame(bitmap: Bitmap) {
        val now = System.currentTimeMillis()
        if (now - lastFrameTimeMs < minFrameIntervalMs) return
        lastFrameTimeMs = now

        val base64 = bitmapToBase64Jpeg(bitmap, quality = 50)
        client.sendFrameToAll(base64)
    }

    /**
     * Compress bitmap to PNG and send as a screenshot result to all desktops.
     */
    fun sendScreenshot(bitmap: Bitmap) {
        val base64 = bitmapToBase64Png(bitmap)
        client.sendScreenshotToAll(base64)
    }

    private fun bitmapToBase64Jpeg(bitmap: Bitmap, quality: Int): String {
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, stream)
        return Base64.encodeToString(stream.toByteArray(), Base64.NO_WRAP)
    }

    private fun bitmapToBase64Png(bitmap: Bitmap): String {
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        return Base64.encodeToString(stream.toByteArray(), Base64.NO_WRAP)
    }
}
